src/backend/metal/src/device.rs

```diff
@@ -887,61 +886,93 @@ impl hal::Device<Backend> for Device {
             let desc = metal::DepthStencilDescriptor::new();
 
             match depth_stencil.depth {
-                DepthTest::On { fun, write } => {
+                pso::DepthTest::On { fun, write } => {
                     desc.set_depth_compare_function(conv::map_compare_function(fun));
                     desc.set_depth_write_enabled(write);
                 }
-                DepthTest::Off => {}
+                pso::DepthTest::Off => {}
             }
 
             match depth_stencil.stencil {
-                StencilTest::On { .. } => {
+                pso::StencilTest::On { .. } => {
                     unimplemented!()
                 }
-                StencilTest::Off => {}
+                pso::StencilTest::Off => {}
             }
 
             device.new_depth_stencil_state(&desc)
         });
 
         // Vertex buffers
-        const STRIDE_GRANULARITY: u32 = 4; //TODO: work around?
         let vertex_descriptor = metal::VertexDescriptor::new();
-        for vertex_buffer in &pipeline_desc.vertex_buffers {
-            let mtl_buffer_index = pipeline_layout.attribute_buffer_index as usize + vertex_buffer.binding as usize;
-            let mtl_buffer_desc = vertex_descriptor
-                .layouts()
-                .object_at(mtl_buffer_index)
-                .expect("too many vertex descriptor layouts");
-            if vertex_buffer.stride % STRIDE_GRANULARITY != 0 {
-                error!("Stride ({}) must be a multiple of {}", vertex_buffer.stride, STRIDE_GRANULARITY);
-                return Err(pso::CreationError::Other);
-            }
-            mtl_buffer_desc.set_stride(vertex_buffer.stride as u64);
-            match vertex_buffer.rate {
-                0 => {
-                    mtl_buffer_desc.set_step_function(MTLVertexStepFunction::PerVertex);
-                }
-                c => {
-                    mtl_buffer_desc.set_step_function(MTLVertexStepFunction::PerInstance);
-                    mtl_buffer_desc.set_step_rate(c as u64);
+        let mut vertex_buffer_map = n::VertexBufferMap::new();
+        let mut next_buffer_index = pipeline_layout.attribute_buffer_index;
+        debug!("Vertex attribute remapping started");
+
+        for (i, &pso::AttributeDesc { binding, element, ..}) in pipeline_desc.attributes.iter().enumerate() {
+            let original = pipeline_desc.vertex_buffers
+                .iter()
+                .find(|vb| vb.binding == binding)
+                .expect("no associated vertex buffer found");
+            // handle wrapping offsets
+            let elem_size = element.format.surface_desc().bits as pso::ElemOffset >> 3;
+            let (cut_offset, base_offset) = if element.offset < original.stride {
+                (element.offset, 0)
+            } else if element.offset + elem_size <= original.stride {
```

Comment:
This case isn't clear to me, shouldn't it be greater than the stride if it's wrapping?